hw/ip/otbn/util/check_loop.py

```diff
@@ -0,0 +1,226 @@
+        if operands['grd'] != 1:
+            to_check.append(pc)
+
+    for pc in to_check:
+        operands = program.get_operands(pc)
+        branch_addr = operands['offset'] & ((1 << 32) - 1)
+
+        # Get the loop bodies the branch is inside, if any
+        current_loops = []
+        for l in loops:
+            if pc in l:
+                current_loops.append(l)
+
+        # Check that we're not branching out of any loop bodies
+        for l in current_loops:
+            if branch_addr not in l:
+                insn = program.get_insn(pc)
+                out.warn(
+                    'Branch out of loop at PC {:#x} (loop from PC {:#x} to PC {:#x}, branch {} to PC {:#x}). This might cause problems with the loop stack and surprising behavior.'
+                    .format(pc, l.start, l.end, insn.mnemonic, branch_addr))
+
+        # Check that we're not branching *into* a loop body that the branch
+        # instruction is not already in
+        for l in loops:
+            if (branch_addr in l) and (l not in current_loops):
+                out.warn(
+                    'Branch into loop at PC {:#x} (loop from PC {:#x} to PC {:#x}, branch {} to PC {:#x}). This might cause problems with the loop stack and surprising behavior.'
+                    .format(pc, l.start, l.end, insn.mnemonic, branch_addr))
+
+    return out
+
+
+def _check_loop_stack(program: OTBNProgram,
+                      loops: List[CodeSection]) -> CheckResult:
+    '''Checks that loops will likely be properly cleared from loop stack.
+
+    The checks here are based on the OTBN hardware IP documentation on loop
+    nesting. From the docs:
+
+        To avoid polluting the loop stack and avoid surprising behaviour, the
+        programmer must ensure that:
+
+        * Even if there are branches and jumps within a loop body, the final
+          instruction of the loop body gets executed exactly once per iteration.
+        * Nested loops have distinct end addresses.
+        * The end instruction of an outer loop is not executed before an inner loop
+          finishes.
+
+    In order to avoid simulating the control flow of the entire program to
+    check the first and third conditions, this check takes a conservative,
+    simplistic approach and simply warns about all branching into or out of
+    loop bodies, including jumps that don't use the call stack (e.g. `jal x0,
+    <addr>`). Branching to locations within the same loop body is permitted.
+
+    The second condition in the list, distinct end addresses, is checked
+    separately.
+    '''
+    out = CheckResult()
+    out += _check_loop_branching(program, loops)
+
+    # Check that loops have unique end addresses
+    end_addrs = []
+    for l in loops:
+        if l.end in end_addrs:
+            out.err(
+                'Loop starting at PC {:#x} shares a final instruction with another loop; consider adding a NOP instruction.'
+                .format(l.start))
```

Comment:
I think this should probably be strengthened to ensure that if two loops intersect then one is completely contained in the other, with an instruction to spare at the end. This would catch things like
```
   loopi x2, 2
      loopi x3, 10
      ...
```
as well.

(Yes, with enthusiastic use of jump instructions you *could* unwind the stack in this case, but I don't think anyone is ever going to do it on purpose!)

Reply:
That's a good point! Let me see about adding this check -- I might do it in a follow-up, though, depending on how complicated it is to check.

Reply:
Implemented!

Reply:
That's clever: much simpler than what I thought you'd have to do!